src/scene/materials/shader-graph-node.js

```diff
@@ -0,0 +1,615 @@
+                    return false;
+                }
+            } else {
+                for (i = 0; i < this.graphData.connections.length; i++) {
+                    if (this.graphData.connections[i]) {
+                        if (!(this.graphData.connections[i].srcVarName && this.graphData.connections[i].dstVarName && (this.graphData.connections[i].srcIndex >= 0 || this.graphData.connections[i].dstIndex >= 0 ))) {
+                            return false;
+                        }
+                    } else {
+                        return false;
+                    }
+                }
+
+                // infinite recursion protection
+                // if no graphRootTime set, set it to current time
+                if (!graphRootTime) {
+                    graphRootTime = new Date().getTime();
+                    this._tmpRootTime = graphRootTime;
+                }
+                for (i = 0; i < this.graphData.subGraphs.length; i++) {
+                    if (this.graphData.subGraphs[i] && this.graphData.subGraphs[i] instanceof ShaderGraphNode && this.graphData.subGraphs[i].name) {
+                        if (this.graphData.subGraphs[i]._tmpRootTime === graphRootTime) {
+                            // recursion detected!
+                            return false;
+                        }
+
+                        this._tmpRootTime = graphRootTime;
+
+                        if (!(this.graphData.subGraphs[i].hasValidGraphData(graphRootTime))) {
+                            return false;
+                        }
+
+                    } else {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+        return false;
+    },
+
+    _addIoPort: function (type, name, value) {
+        var ioPort;
+        if (value instanceof Texture) {
+            ioPort = { type: type, name: name, valueTex: value };
+        } else if (value instanceof Vec4) {
+            ioPort = { type: type, name: name, valueX: value.x, valueY: value.y, valueZ: value.z, valueW: value.w };
+        } else if (value instanceof Vec3) {
+            ioPort = { type: type, name: name, valueX: value.x, valueY: value.y, valueZ: value.z };
+        } else if (value instanceof Vec2) {
+            ioPort = { type: type, name: name, valueX: value.x, valueY: value.y };
+        } else if (typeof(value) === 'number') {
+            ioPort = { type: type, name: name, valueX: value };
+        } else {
+            ioPort = { type: type, name: name }; // output ports do not require a value
+        }
+
+        this.graphData.ioPorts.push(ioPort);
+
+        return ioPort;
+    },
+
+    addInput: function (type, name, value) {
+        return this._addIoPort(type, 'IN_' + name, value);
+    },
+
+    addOutput: function (type, name, value) {
+        return this._addIoPort(type, 'OUT_' + name, value);
+    },
+
+    addConstant: function (type, value) {

```

Comment:
curious if this could take [optional] name parameter to be used to make generated code more readable? (to be used instead of 'this.graphData.ioPorts.length' when supplied)

Reply:
yes - I can add that

Reply:
I added an optional userId - from past experience building material graphs, its rare to name constants - unless you convert it to a parameter - but this could be useful for debugging